### PR TITLE
misc: add Containerfile

### PR DESCRIPTION
**Pull request by mvo5**
This PR adds Containerfile integration, with that a bib style deploy/run is possible. The Containerfile is pretty nasty right now, especially around the "getting supported repositories" part. This needs to get fixed either in composer or images IMHO (or via go:embed). But with that we can give it to our early users and let them play with building images.

It also contains automatic container building as part of the GH action (thanks @supakeen!) and once we have it we can update the README to point to the auto-generated containers for even easier testing/using.

As is "list-images" and "manifest" will work but once #8 is merged the below works (on my local system, we want tests here :)

Run build via container, bib-style:
```console
$ sudo podman build . -t test-icbli
$ sudo podman run --privileged \
   -v ./output:/output \
   test-ibcli:latest \
   build \
   --distro fedora-41 \
   minimal-raw
```

### CODE REVIEW

**Comment by supakeen** [comment]
I've been rude and committed CI for the container build, we might need to change organisation or repository permissions for it to upload.

**Review by supakeen — APPROVED** [review]
Looks good (though I guess I wrote part of it so maybe wait for a second review).

File: Containerfile
Comment (ondrejbudai) [inline Containerfile:13]: Can we put these in a single RUN command, and clean the dnf cache afterwards, to save some disk space and reduce the number of layers?

```
RUN dnf install -y dnf-plugins-core \
  && dnf copr enable -y @osbuild/osbuild \
  && dnf install -y libxcrypt-compat wget osbuild osbuild-ostree osbuild-depsolve-dnf osbuild-lvm2 \
  && dnf clean all
```

File: README.md
Comment (ondrejbudai) [inline README.md:7]: Leftover TODO

File: entrypoint.sh
Comment (ondrejbudai) [inline entrypoint.sh:12]: I think this is currently unused, and I think we should actually use a "real" filesystem for the osbuild store, because tmpfs might not be big enough for image building.

File: Containerfile
Comment (ondrejbudai) [inline Containerfile:52]: I think this is currently unused.

File: Containerfile
Comment (ondrejbudai) [inline Containerfile:51]: This is currently unused, and should be used (osbuild's store cannot be on on overlayfs). How can we pass the `/store` path to image builder cli to utilize it for the osbuild's store?

File: .github/workflows/container.yaml
Comment (ondrejbudai) [inline .github/workflows/container.yaml:6]: Also build in a merge queue?

File: Containerfile
Comment (supakeen) [reply Containerfile]: I've addressed this in a separate commit.

File: README.md
Comment (supakeen) [reply README.md]: I've addressed this in a separate commit.

**Comment by supakeen** [comment]
I've made 3 separate commits to take care of the low hanging fruit things, feel free to reorganize/rebase how you see fit @mvo5 

(I don't want to force push to your branch at all).

The other remarks will take a bit more thinking.

File: entrypoint.sh
Comment (mvo5) [reply entrypoint.sh]: Thanks, indeed, I removed this and followed our pattern from bib and added `--store` and use that in the entrypoint.

File: Containerfile
Comment (mvo5) [reply Containerfile]: Indeed, I removed it (for now, I think we want to add this later again once we grow a --rpmmd like flag).

**Comment by mvo5** [comment]
> I've made 3 separate commits to take care of the low hanging fruit things, feel free to reorganize/rebase how you see fit @mvo5
> 
> (I don't want to force push to your branch at all).
> 
> The other remarks will take a bit more thinking.

Thank you! This is really appreciated! I squashed them into their relevant places (hopefully :) and also addressed the points from @ondrejbudai - I think this is ready for a re-review. I can split out further if needed but it seems it's not toooo big .

**Review by achilleas-k — APPROVED** [review]
LGTM.

**Comment by achilleas-k** [comment]
One thought about the new `--store` flag (to consider in a follow-up):
Would it make sense to separate osbuild flags into their own group?  Could we have an `--osbuild-opts` flag for example that can take options to be passed down to osbuild?

So we could do:
```
image-builder-cli ... --osbuild-opts="store=/data/osbuild-store,checkpoint=build,checkpoint=os,cache-max-size='2 GiB'" ...
```

It's not the prettiest but it saves us from having to support every osbuild flag.  OTOH, it also forces users to look up osbuild flags if they want to discover everything that's supported.

If we don't like this, I think eventually we should support every option that osbuild supports (except maybe the `--monitor` option).

File: .github/workflows/go.yml
Comment (ondrejbudai) [inline .github/workflows/go.yml:16]: 😍

File: test/containerbuild.py
Comment (ondrejbudai) [inline test/containerbuild.py:1]: I think we just need `subprocess` and `pytest` here.

File: test/test_container.py
Comment (ondrejbudai) [inline test/test_container.py:1]: Would you be willing to use Go for integration tests? Is there any reason to actually use Python?

I can tell you why I prefer Go:
- developers don't need to switch their language mindset when writing integration tests
- we don't need an extra linter and formatter
- sharing code between integration tests and the SUT
- pytest fixtures are cool... until they start being overwhelming. I think I have a preference for the lightweight nature of Go.

File: Containerfile
Comment (ondrejbudai) [inline Containerfile:52]: I had an idea: image-builder-cli is supposed to be an RPM in the future. So this path should be FHS-compliant, and ideally it should be the same between the container image and the RPM.

Something like `/var/cache/image-builder/{store,dnf-cache}`?

Note that I have no concerns about `/output`, this is perfectly valid to be container-specific.

File: test/test_container.py
Comment (ondrejbudai) [inline test/test_container.py:20]: Can we do something quick and dirty to pave over the RPM downloading issue that our stack is suffering from? Override repos with rpmrepo.osbuild.org for the test?

File: .github/workflows/pytest.yml
Comment (ondrejbudai) [inline .github/workflows/pytest.yml:7]: no merge queue here

File: Containerfile
Comment (mvo5) [reply Containerfile:52]: Love it! will do in a followup

File: test/test_container.py
Comment (mvo5) [reply test/test_container.py:20]: I will try centos-9 instead, from my experience with bib it seems the centos mirrors are generally a bit more stable.

File: test/containerbuild.py
Comment (mvo5) [reply test/containerbuild.py:1]: Yeah, silly me - and a linter :)

File: test/test_container.py
Comment (mvo5) [reply test/test_container.py:1]: Sure, I will give this a "go" (see what I did here ;)

**Pull request merged**
